docs/proposals/high-availability.md

```diff
@@ -0,0 +1,34 @@
+# High Availability of Daemons in Kubernetes
+This document serves as a proposal for high availability of the master daemons in kubernetes.
+
+## Design Options
+1. Hot Standby Daemons: In this scenario, data and state are shared between the two deamons such that an immediate failure in one daemon causes the the standby deamon to take over exactly where the failed daemon had left off.  This would be an ideal solution for kubernetes, however it poses a series of challenges in the case of controllers where daemon-state is cached locally and not persisted in a transactional way to a storage facility.  As a result, we are **NOT** planning on this approach. 
```

Comment:
Would be great to give example(s) of the kind of local non-persistent state the controllers have.

Also, note that there's a hybrid between (1) and (2), which is actually what we do for the standby replicas in Borg. The standby replicas keep a local copy of the etcd state (for example, have open watches on objects they would manage if they were to take over, and update their local data structures on updates), and then when one gets the signal that it is the new master, it does a set of operations to build the non-persisted state that it will need, before it starts serving. You can look at this as a hot standby that does some additional work before it can become the new master, or a cold standby that gets a head-start on becoming the new master by tracking all the changes to persistent state.

Oh, actually, maybe this *is* what you intended by cold standby? I was thinking of cold standby as equivalent to just having a machine with resources guaranteed to be available, but no process running until it is needed.